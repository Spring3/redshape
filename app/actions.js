export const LOGIN = 'LOGIN';
const login = data => ({ type: LOGIN, data });

export const LOGOUT = 'LOGOUT';
const logout = () => ({ type: LOGOUT });

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';
const updateSettings = data => ({ type: UPDATE_SETTINGS, data });

export default {
  user: {
    login,
    logout
  },
  settings: {
    update: updateSettings
  }
};
