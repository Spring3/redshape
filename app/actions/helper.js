import * as axios from '../../modules/request';

function IssueFilter () {
  const filter = {};

  this.issue = (ids) => {
    filter.issue_id = Array.isArray(ids) ? ids.join(',') : ids;
    return this;
  };

  this.project = (id) => {
    filter.project_id = id;
    return this;
  };

  this.subproject = (id) => {
    filter.subprojectId = id;
    return this;
  };

  this.tracker = (id) => {
    filter.tracker_id = id;
    return this;
  };

  this.status = ({ open, closed }) => {
    if (open && closed) {
      filter.status_id = '*';
    } else {
      filter.status_id = open ? 'open' : 'closed';
    }
    return this;
  };

  this.assignee = (id) => {
    filter.assigned_to_id = id;
    return this;
  };

  this.author = (id) => {
    filter.author_id = id;
    return this;
  };

  this.title = (str) => {
    if (str) {
      filter.subject = encodeURIComponent(`~${str.trim()}`);
    }
    return this;
  };

  this.due = (date) => {
    filter.due_date = date.toISOString();
    return this;
  };

  this.createdAt = (date) => {
    filter.created_on = date.toISOString();
    return this;
  };

  this.createdBetween = (startDate, endDate) => {
    filter.created_on = encodeURIComponent(`><${startDate.toISOString()}|${endDate.toISOString()}`);
    return this;
  };

  this.priority = (id) => {
    filter.priority_id = id;
    return this;
  };

  this.sort = (by, direction) => {
    if (by && direction) {
      const column = by.indexOf('.') !== -1
        ? by.substring(0, by.indexOf('.'))
        : by;
      filter.sort = `${column}:${direction}`;
    }
    return this;
  };

  this.build = () => ({ ...filter });
}

const handleReject = (e) => {
  console.log('EEEE', e);
  // if this request was not cancelled
  if (!axios.default.isCancel(e)) {
    return Promise.reject(new Error(`Error ${e.status} (${e.statusText || e.message})`));
  }
  console.log('Request was cancelled');
  return undefined;
};

const request = ({
  url,
  method = 'GET',
  data,
  headers,
  query,
  id
}) => {
  const requestConfig = {
    url,
    method,
    headers: axios.makeCancellable(headers || {}, id)
  };

  if (!['GET', 'DELETE'].includes(method)) {
    requestConfig.data = typeof data === 'string' ? data : JSON.stringify(data);
  }

  if (query) {
    requestConfig.params = query;
  }

  if (!axios.getInstance()) {
    return Promise.reject(new Error('401 - Unauthorized'));
  }

  return axios.getInstance().request(requestConfig)
    .then((res) => {
      delete axios.pendingRequests[id];
      return ({ data: res.data });
    })
    .catch((error) => {
      delete axios.pendingRequests[id];
      return handleReject(error);
    });
};

const login = ({
  redmineEndpoint,
  url,
  headers
}) => {
  const defaultConfig = axios.getDefaultConfig();
  return axios.default.request({
    baseURL: redmineEndpoint,
    timeout: defaultConfig.timeout,
    headers: {
      ...defaultConfig.headers,
      ...(headers || {})
    },
    url,
    method: 'GET'
  }).then((res) => {
    axios.reset();
    const { api_key } = res.data.user || {};
    if (api_key) {
      axios.initialize(redmineEndpoint, api_key);
    }
    return { data: res.data };
  });
};

const notify = {
  start: (type, id) => ({ type, status: 'START', id }),
  paginate: (type, data, id) => ({ type, data, status: 'PAGE_NEXT', id }),
  ok: (type, data, id) => ({ type, data, status: 'OK', id }),
  nok: (type, data, id) => ({ type, data, status: 'NOK', id }),
  cancel: type => ({ type, status: 'CANCELLED' })
};

export {
  IssueFilter,
  notify,
  login
};

export default request;
