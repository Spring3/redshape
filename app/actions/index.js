import userActions from './user.actions';
import trackingActions from './tracking.actions';
import issuesActions from './issues.actions';
import projectActions from './project.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  user: userActions,
  issues: issuesActions,
  tracking: trackingActions,
  projects: projectActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
