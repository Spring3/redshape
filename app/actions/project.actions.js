import request, { notify } from './helper';

export const PROJECT_GET_ALL = 'PROJECT_GET_ALL';

const getAll = () => async (dispatch) => {
  dispatch(notify.start(PROJECT_GET_ALL));

  const requestPage = (offset = 0) => request({
    url: '/projects.json',
    query: {
      include: 'time_entry_activities',
      offset
    },
    id: `getProjects:${offset}`
  }).then(({ data }) => data);

  return requestPage()
    .then((data) => {
      const { total_count, offset, limit } = data;
      if (total_count > (offset + limit)) {
        // Math.floor because we already fetched the 1st page.
        const remainingPages = Math.floor(total_count / limit);
        const promises = [data];
        for (let i = 0; i < remainingPages; i++) {
          promises.push(requestPage((i + 1) * limit));
        }
        return Promise.all(promises);
      }
      return [data];
    })
    .then((results) => {
      dispatch(notify.ok(PROJECT_GET_ALL, results));
    })
    .catch((error) => {
      console.error('Error when trying to get the info about the existing projects', error);
      dispatch(notify.nok(PROJECT_GET_ALL, error));
    });
};

export default {
  getAll
};
