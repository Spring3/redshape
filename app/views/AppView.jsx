import React, { Fragment } from 'react';

export default () => (<Fragment/>);
