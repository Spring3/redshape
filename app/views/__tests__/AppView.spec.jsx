import React from 'react';
import { HashRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render, fireEvent, cleanup, wait } from 'react-testing-library';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { USER_LOGOUT } from '../../actions/user.actions';

import AppView from '../AppView';

const mockStore = configureStore([thunk]);

describe('AppView', () => {
  afterEach(() => {
    cleanup();
    fetch.resetMocks();
  });

  it('should match the snapshot', () => {
    const store = mockStore({
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      }
    });
    const tree = renderer.create(
      <Provider store={store}>
        <HashRouter><AppView /></HashRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('[integration] AppView', () => {
    it('should open / close the side bar', () => {
      fetch.mockResponse(JSON.stringify({ ok: false, status: 400, statusText: 'Bad Request' }));
      const store = mockStore({
        user: {
          id: 1,
          firstname: 'firstname',
          lastname: 'lastname',
          redmineEndpoint: 'https://redmine.domain',
          api_key: '123abc'
        },
        issues: {
          all: {
            data: []
          }
        }
      });
      const wrapper = mount(
        <Provider store={store}>
          <HashRouter>
            <AppView />
          </HashRouter>
        </Provider>
      );

      let view = wrapper.find('AppView');
      expect(view.exists()).toBeTruthy();
      expect(view.state('showSidebar')).toBe(false);
      expect(view.state('showFooter')).toBe(false);

      let aside = wrapper.find('aside');
      expect(aside.exists()).toBeTruthy();
      expect(aside).toHaveStyleRule('display', 'none');

      let hamburgerButton = wrapper.find('li#hamburger').childAt(0);
      expect(hamburgerButton.exists()).toBeTruthy();

      hamburgerButton.simulate('click');
      wrapper.update();

      view = wrapper.find('AppView');
      aside = wrapper.find('aside');
      hamburgerButton = wrapper.find('li#hamburger').childAt(0);

      expect(aside).not.toHaveStyleRule('display', 'none');
      expect(view.state('showSidebar')).toBe(true);

      hamburgerButton.simulate('click');
      wrapper.update();

      view = wrapper.find('AppView');
      aside = wrapper.find('aside');
      expect(aside).toHaveStyleRule('display', 'none');
      expect(view.state('showSidebar')).toBe(false);
    });

    it('should have a button to log out', () => {
      fetch.mockResponse(JSON.stringify({ ok: false, status: 400, statusText: 'Bad Request' }));
      const store = mockStore({
        user: {
          id: 1,
          firstname: 'firstname',
          lastname: 'lastname',
          redmineEndpoint: 'https://redmine.domain',
          api_key: '123abc'
        },
        issues: {
          all: {
            data: []
          }
        }
      });

      const historyMock = {
        location: {
          pathname: '/app'
        },
        listen: () => () => {},
        push: jest.fn()
      };

      const wrapper = mount(
        <Provider store={store}>
          <Router history={historyMock}>
            <AppView />
          </Router>
        </Provider>
      );

      const signoutButton = wrapper.find('#signout').hostNodes();
      expect(signoutButton.exists()).toBeTruthy();

      signoutButton.simulate('click');
      wrapper.update();

      expect(historyMock.push).toHaveBeenCalledWith('/');
      const actions = store.getActions();
      expect(actions.length).toBeGreaterThan(0);
      expect(actions.pop()).toEqual({ type: USER_LOGOUT });
    });
  });
});
