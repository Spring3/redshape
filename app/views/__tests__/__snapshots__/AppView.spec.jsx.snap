// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView should match the snapshot 1`] = `
.c3 {
  display: block;
  width: 100%;
  border-radius: 5px;
  padding: 5px 10px;
  box-sizing: border-box;
  font-size: 14px;
  border: 1px solid #FF7079;
  min-height: 35px;
  outline: none;
  color: #FF7079;
  font-weight: bold;
  background: #FAFAFA;
}

.c3:hover {
  border: 1px solid #FF7079;
}

.c3:focus {
  box-shadow: 0px 0px 0px 1px #FF7079;
}

.c13 h4 {
  margin-bottom: 10px;
  color: #A4A4A4;
}

.c9 {
  padding: 10px 5px;
  border-radius: 5px;
  border: 2px solid transparent;
  box-shadow: 0px 0px 0px 2px #FF7079;
  color: #FF7079;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  width: auto;
  margin: 25px auto 0px auto;
}

.c9:hover,
.c9:focus {
  cursor: pointer;
  background: #FF7079;
  border: 2px solid #FF7079;
  box-shadow: 0px 0px 0px 2px #EF6069;
  color: white;
}

.c9:active {
  background: #EF6069;
}

.c2 {
  outline: none;
}

.c2:active,
.c2:focus,
.c2:visited {
  background: transparent;
}

.c16 tbody {
  text-align: center;
  font-size: 14px;
  overflow-y: scroll;
  display: block;
  max-height: 450px;
}

.c16 tbody tr th {
  padding: 5px;
  border: 2px solid transparent;
  background: white;
}

.c16 tbody tr th:hover {
  cursor: pointer;
  background: #FF7079;
  color: white;
}

.c16 tbody tr th:last-child {
  width: 90px;
}

.c16 tbody tr:hover {
  cursor: pointer;
}

.c16 tbody tr:hover td {
  background: #FF7079;
  color: white;
}

.c16 tbody tr:nth-child(even) {
  background: #F2F2F2;
}

.c11 {
  display: grid;
  padding: 20px;
  grid-template-columns: repeat(auto-fit,minmax(100px,1fr));
  grid-auto-rows: 100px;
  grid-auto-flow: dense;
  grid-gap: 20px;
}

.c12 {
  background: white;
  padding: 20px;
  border-radius: 5px;
  grid-column: span 8;
  grid-row: span 6;
}

.c17 {
  background: white;
  padding: 20px;
  border-radius: 5px;
  grid-column: span 3;
  grid-row: span 2;
}

.c18 {
  background: white;
  padding: 20px;
  border-radius: 5px;
  grid-column: span 3;
  grid-row: span 4;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c15 {
  margin: 20px 10px 20px 0px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  height: 100%;
  display: grid;
  grid-template-rows: 60px auto;
  grid-template-columns: repeat(auto-fit,minmax(70px,1fr));
}

.c4 {
  border-top: 3px solid #FF7079;
  grid-row: 1;
  grid-column: span 12;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0px 20px;
}

.c1 {
  display: none;
  background: tomato;
  grid-row: 1 / -1;
  grid-column: span 3;
  padding-top: 25px;
}

.c19 {
  display: none;
  background: sandybrown;
}

.c10 {
  background: aliceblue;
  grid-column: span 12;
  grid-row: 2 / -1;
}

.c5 {
  list-style-type: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 li {
  display: inline;
  margin: 0px 10px;
  font-size: 14px;
}

.c5 li:hover {
  cursor: pointer;
}

.c5 li:first-child {
  margin-left: 0;
}

.c7 {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.c7 li {
  display: inline;
}

.c7 li:first-child {
  margin-right: 10px;
}

.c8 {
  margin: 0;
  padding: 5px;
  font-size: 14px;
}

.c6 {
  padding: 0px;
  margin: 0px;
}

.c6 svg {
  vertical-align: middle;
}

<div
  className="c0"
>
  <aside
    className="c1"
  >
    <div>
      <a
        className="c2"
        disabled={false}
        href="#"
        onClick={[Function]}
      >
        <svg
          className="mdi-icon "
          fill="currentColor"
          height={24}
          viewBox="0 0 24 24"
          width={24}
        >
          <path
            d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
          />
        </svg>
      </a>
      <h4>
        Time Log History
      </h4>
    </div>
    <div>
      <input
        className="c3"
        disabled={false}
        onChange={[Function]}
        placeholder="search"
        type="text"
      />
    </div>
    <ul>
      <li>
        <span>
          Entry Name
        </span>
        <span>
          1.0 hours
        </span>
        <span>
          1 day(s) ago
        </span>
        <button>
          Edit
        </button>
      </li>
      <li>
        <span>
          Entry Name
        </span>
        <span>
          1.5 hours
        </span>
        <span>
          1 day(s) ago
        </span>
        <button>
          Edit
        </button>
      </li>
      <li>
        <span>
          Entry Name
        </span>
        <span>
          1.3 hours
        </span>
        <span>
          3 day(s) ago
        </span>
        <button>
          Edit
        </button>
      </li>
    </ul>
  </aside>
  <nav
    className="c4"
  >
    <ul
      className="c5"
    >
      <li
        className="c6"
        id="hamburger"
      >
        <a
          className="c2"
          disabled={false}
          href="#"
          onClick={[Function]}
        >
          <svg
            className="mdi-icon "
            fill="currentColor"
            height={24}
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              d="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z"
            />
          </svg>
        </a>
      </li>
      <li>
        My Page
      </li>
      <li>
        Time
      </li>
      <li>
        Issues
      </li>
    </ul>
    <ul
      className="c7"
    >
      <li>
        firstname
         
        lastname
      </li>
      <li>
        <button
          className="c8 c9"
          disabled={false}
          id="signout"
          onClick={[Function]}
          type="button"
        >
          Sign out
        </button>
      </li>
    </ul>
  </nav>
  <div
    className="c10"
  >
    <div
      className="c11"
    >
      <section
        className="c12"
      >
        <h2>
          Issues assigned to me
        </h2>
        <div>
          <div
            className="form-group undefined c13"
          >
            <h4
              htmlFor="headers"
            >
              Table Columns
            </h4>
            <div
              className="css-10nd86i"
              onKeyDown={[Function]}
            >
              <div
                className="css-vj8t7z"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className="css-1hwfws3"
                >
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Id
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Subject
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Project
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Tracker
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Status
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Priority
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Estimation
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    style={
                      Object {
                        "overflow": "hidden",
                        "whiteSpace": "nowrap",
                        "width": "auto",
                      }
                    }
                  >
                    <div
                      className="css-xwjg1b"
                    >
                      <div
                        className="css-12jo7m5"
                      >
                        Due Date
                      </div>
                      <div
                        className="css-1alnv5e"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-19bqh2r"
                          focusable="false"
                          height={14}
                          viewBox="0 0 20 20"
                          width={14}
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    className="css-1g6gooi"
                  >
                    <div
                      className=""
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-autocomplete="list"
                        autoCapitalize="none"
                        autoComplete="off"
                        autoCorrect="off"
                        disabled={false}
                        id="react-select-2-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        spellCheck="false"
                        style={
                          Object {
                            "background": 0,
                            "border": 0,
                            "boxSizing": "content-box",
                            "color": "inherit",
                            "fontSize": "inherit",
                            "opacity": 1,
                            "outline": 0,
                            "padding": 0,
                            "width": "1px",
                          }
                        }
                        tabIndex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="css-1wy0on6"
                >
                  <div
                    aria-hidden="true"
                    className="css-1ep9fjw"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-19bqh2r"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                      />
                    </svg>
                  </div>
                  <span
                    className="css-d8oujb"
                  />
                  <div
                    aria-hidden="true"
                    className="css-1ep9fjw"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-19bqh2r"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div>
                <input
                  name="headers"
                  type="hidden"
                  value="id"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="subject"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="project"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="tracker"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="status"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="priority"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="estimation"
                />
                <input
                  name="headers"
                  type="hidden"
                  value="dueDate"
                />
              </div>
            </div>
          </div>
          <div
            className="c14"
          >
            <div
              className="c15"
            >
              <input
                className="c3"
                disabled={false}
                name="search"
                onChange={[Function]}
                placeholder="Search"
                type="text"
              />
            </div>
            <div>
              <div
                className="form-group undefined c13"
              >
                <input
                  checked={false}
                  className=""
                  disabled={false}
                  name="showClosed"
                  onChange={[Function]}
                  type="checkbox"
                />
                <label
                  htmlFor="showClosed"
                >
                  Show Closed
                </label>
              </div>
            </div>
          </div>
        </div>
        <table
          cellPadding={10}
          className="c16"
          width="100%"
        >
          <tbody>
            <tr>
              <th>
                Id
              </th>
              <th>
                Subject
              </th>
              <th>
                Project
              </th>
              <th>
                Tracker
              </th>
              <th>
                Status
              </th>
              <th>
                Priority
              </th>
              <th>
                Estimation
              </th>
              <th>
                Due Date
              </th>
            </tr>
          </tbody>
        </table>
      </section>
      <section
        className="c17"
      >
        <h2>
          Time spent
        </h2>
      </section>
      <section
        className="c18"
      >
        <h2>
          Activity Stack
        </h2>
      </section>
    </div>
    <div
      className="c19"
    >
      <span>
        Task name
      </span>
      <span>
        00:00:00
      </span>
      <button>
        Stop
      </button>
    </div>
  </div>
</div>
`;
